File: LeetCode/713/main.cpp
class Solution {
public:
  int numSubarrayProductLessThanK(vector<int> &arr, int k) {

    int n = arr.size();
    int left = 0, prod = 1, ans = 0;

    for (int i = 0; i < n; i++) {
      prod *= arr[i];
      while (left <= i && prod >= k) {
        prod /= arr[left];
        left++;
      }
      ans += (i - left + 1);
    }
    return ans;
  }
};

File: LeetCode/713/main.py
class Solution:
    def numSubarrayProductLessThanK(self, arr: List[int], k: int) -> int:
        n = len(arr)
        left, prod, ans = 0, 1, 0
        for i in range(n):
            prod *= arr[i]
            while left <= i and prod >= k:
                prod //= arr[left]
                left += 1

            ans += i - left + 1

        return ans
